src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -70,3 +111,203 @@ def dss_deletion_subscription_query(self, prefix: str) -> JSON:
+                        entities[entity_ref[category + '_type']].add(entity_ref[category + '_id'])
+            elif link_type == 'supplementary_file_link':
+                # For MVP, only project entities can have associated supplementary files.
+                associate_type = link['entity']['entity_type']
+                associate_id = link['entity']['entity_id']
+                require(associate_type == 'project',
+                        f'Supplementary file must be associated with entity of type "project", '
+                        f'not "{associate_type}"')
+                require(associate_id == bundle_project_id,
+                        f'Supplementary file must be associated with the current project '
+                        f'({bundle_project_id}, not {associate_id})')
+                for supp_file in link['files']:
+                    entities['supplementary_file'].add(supp_file['file_id'])
+            else:
+                raise RequirementError(f'Unexpected link_type: {link_type}')
+
+        def retrieve_rows(entity_type: str, entity_ids: Set[str]):
+            pk_column = entity_type + '_id'
+            non_pk_columns = bundle.data_columns if entity_type.endswith('_file') else bundle.metadata_columns
+            columns = ', '.join(non_pk_columns | {pk_column})
+            uuid_in_list = ' OR '.join(f'{pk_column} = "{entity_id}"' for entity_id in entity_ids)
+            rows = self._query_latest_version(f'''
+                SELECT {columns}
+                FROM {self._source.bq_name}.{entity_type}
+                WHERE {uuid_in_list}
+            ''', group_by=pk_column)
+            log.info('Retrieved %s %s entities', len(rows), entity_type)
+            return rows
+
+        with ThreadPoolExecutor(max_workers=config.num_repo_workers) as executor:
+            futures = {
+                entity_type: executor.submit(retrieve_rows, entity_type, entity_ids)
+                for entity_type, entity_ids in entities.items()
+            }
+        for entity_type, future in futures.items():
+            rows = future.result()
+            entity_ids = entities[entity_type]
+            for i, row in enumerate(rows):
+                bundle.add_entity(f'{entity_type}_{i}.json', entity_type, row)
+                entity_ids.remove(row[entity_type + '_id'])
+            if entity_ids:
+                raise RuntimeError(f'Required entities not found in {self._source.bq_name}.{entity_type}: {entity_ids}')
+        return bundle
+
+
+class Checksums(NamedTuple):
+    crc32c: str
+    sha1: str
+    sha256: str
+    s3_etag: str
+
+    def asdict(self) -> Dict[str, str]:
+        return self._asdict()
+
+    @classmethod
+    def without_values(cls) -> Dict[str, str]:
+        return {f: None for f in cls._fields}
+
+    @classmethod
+    def extract(cls, json: JSON) -> 'Checksums':
+        return cls(**{f: json[f] for f in cls._fields})
+
+
+class ManifestEntry(NamedTuple):
```

Comment:
Could you explain the motivation for the existence of this class? It closely resembles the class of the same name in the metadata API and I only see very short-lived instances that are created and then immediately converted to a dict again. Both smell fishy to me.

Reply:
I don't recall the original motivation. Looking at it now I think it would better be replaced with a private method in the `Bundle` class.